3rdparty/find_dependencies.cmake

```diff
@@ -0,0 +1,874 @@
+        else()
+            message(STATUS "Unable to find installed googletest")
+            set(BUILD_GOOGLETEST ON)
+        endif()
+    endif()
+    if(BUILD_GOOGLETEST)
+        build_3rdparty_library(3rdparty_googletest DIRECTORY googletest
+            SOURCES
+                googletest/src/gtest-all.cc
+                googlemock/src/gmock-all.cc
+            INCLUDE_DIRS
+                googletest/include/
+                googletest/
+                googlemock/include/
+                googlemock/
+        )
+        set(GOOGLETEST_TARGET "3rdparty_googletest")
+    endif()
+endif()
+
+# Headless rendering
+if (ENABLE_HEADLESS_RENDERING)
+    find_package(OSMesa REQUIRED)
+    add_library(3rdparty_osmesa INTERFACE)
+    target_include_directories(3rdparty_osmesa INTERFACE ${OSMESA_INCLUDE_DIR})
+    target_link_libraries(3rdparty_osmesa INTERFACE ${OSMESA_LIBRARY})
+    if(NOT BUILD_SHARED_LIBS)
+        install(TARGETS 3rdparty_osmesa EXPORT ${PROJECT_NAME}Targets
+        RUNTIME DESTINATION ${Open3D_INSTALL_BIN_DIR}
+        ARCHIVE DESTINATION ${Open3D_INSTALL_LIB_DIR}
+        LIBRARY DESTINATION ${Open3D_INSTALL_LIB_DIR}
+    )
+    endif()
+    set(OPENGL_TARGET "3rdparty_osmesa")
+    list(APPEND Open3D_3RDPARTY_PRIVATE_TARGETS "${OPENGL_TARGET}")
+else()
+    find_package(OpenGL)
+    if(TARGET OpenGL::GL)
+        if(NOT BUILD_SHARED_LIBS)
+            list(APPEND Open3D_3RDPARTY_EXTERNAL_MODULES "OpenGL")
+        endif()
+        set(OPENGL_TARGET "OpenGL::GL")
+        list(APPEND Open3D_3RDPARTY_PRIVATE_TARGETS "${OPENGL_TARGET}")
+    endif()
+endif()
+
+# imgui
+if(ENABLE_GUI)
+    if(NOT BUILD_IMGUI)
+        find_package(ImGui)
+        if(TARGET ImGui::ImGui)
+            message(STATUS "Using installed third-party library ImGui")
+            if(NOT BUILD_SHARED_LIBS)
+                list(APPEND Open3D_3RDPARTY_EXTERNAL_MODULES "ImGui")
+            endif()
+            set(IMGUI_TARGET "ImGui::ImGui")
+        else()
+            message(STATUS "Unable to find installed third-party library ImGui")
+            set(BUILD_IMGUI ON)
+        endif()
+    endif()
+    if(BUILD_IMGUI)
+        build_3rdparty_library(3rdparty_imgui DIRECTORY imgui
+            SOURCES
+                imgui_demo.cpp
+                imgui_draw.cpp
+                imgui_widgets.cpp
+                imgui.cpp
+        )
+        set(IMGUI_TARGET "3rdparty_imgui")
+    endif()
+    list(APPEND Open3D_3RDPARTY_PRIVATE_TARGETS "${IMGUI_TARGET}")
+endif()
+
+# Filament
+if(ENABLE_GUI)
+    if(BUILD_FILAMENT)
+        message(STATUS "Building third-party library Filament from source")
+        if(MSVC OR (CMAKE_C_COMPILER_ID STREQUAL "Clang" AND CMAKE_CXX_COMPILER_ID STREQUAL "Clang"))
+            set(FILAMENT_C_COMPILER "${CMAKE_C_COMPILER}")
+            set(FILAMENT_CXX_COMPILER "${CMAKE_CXX_COMPILER}")
+        else()
+            find_program(CLANG_CC NAMES clang-7 clang-8 clang-9 clang-10 clang)
```

Comment:
Q2: Shall we really allow falling back to `clang`? In my case, I have `clang-6` installed only, and the compilation fails in the middle without proper indication that the compiler is not supported. Maybe we can call `clang --version` and extract the version?

Reply:
Sure, we can do that. Come to think of it, I should probably switch the order so the newest available compiler version is used.

Reply:
This might be useful: https://stackoverflow.com/a/16209709/1255535